File: examples/python/output/table_advanced.py
#!/usr/bin/env python3
"""Tables: colspan/rowspan, stripes, per-row background and markup cells.

    make run-example EX=python/output/table_advanced
"""

from __future__ import annotations

import sparcli as sc


def print_spans_table() -> None:
    """Colspan/rowspan via Cell; covered slots use Cell.skip / Cell.row_skip."""
    table = sc.Table()
    table.column('Quarter').column('Region') \
         .column('Revenue', sc.ColOpts(halign=sc.Align.RIGHT))

    # Rowspan: "Q1" covers two rows; the covered slot uses Cell.row_skip().
    table.row([sc.Cell('Q1', rowspan=2), 'Europe', '1.2 M'])
    table.row([sc.Cell.row_skip(), 'Americas', '2.4 M'])
    # Colspan: one cell spanning the last two columns.
    table.row(['Q2', sc.Cell('no data yet', colspan=2, halign=sc.Align.CENTER),
               sc.Cell.skip()])

    table.print(sc.TableOpts(border=sc.BorderType.SINGLE, header_row=True,
                             header_style=sc.Style.bold(), title='Spans'))


def print_styled_table() -> None:
    """Stripes, a header column and markup cells; per-row background."""
    table = sc.Table()
    table.column('Check').column('Result') \
         .column('Duration', sc.ColOpts(halign=sc.Align.RIGHT))

    table.row(['build', sc.Cell.markup('[green]✔ passed[/]'), '41 s'])
    table.row(['lint', sc.Cell.markup('[green]✔ passed[/]'), '3 s'])
    # Per-row background highlights the failing check.
    table.row(['fuzz', sc.Cell.markup('[red]✖ failed[/]'), '122 s'],
              bg=sc.Color.rgb(60, 30, 30))

    table.print(sc.TableOpts(border=sc.BorderType.ROUNDED, header_col=True,
                             header_col_bg=sc.Color.rgb(40, 40, 60),
                             striped=True, stripe_bg=sc.Color.rgb(30, 30, 30),
                             cell_pad=sc.Edges(right=1, left=1)))


def main() -> None:
    print_spans_table()
    sc.println('')
    print_styled_table()


if __name__ == '__main__':
    main()

File: examples/rust/output/table_advanced.rs
//! Tables: colspan, stripes, per-row background and markup cells.
//!
//!   cargo run -p sparcli --example output_table_advanced

use sparcli::*;

fn main() {
    print_spans_table();
    println("", Style::default());
    print_styled_table();
}

/// Colspan via `Cell::colspan`; the covered slot uses `Cell::skip`.
fn print_spans_table() {
    let mut table = Table::new();
    table
        .column("Quarter", ColOpts::new())
        .column("Region", ColOpts::new())
        .column("Revenue", ColOpts::new().align(Align::Right));

    table.row(["Q1", "Europe", "1.2 M"]);
    table.row(["Q1", "Americas", "2.4 M"]);
    table.row([
        Cell::new("Q2"),
        Cell::new("no data yet").colspan(2).align(Align::Center),
        Cell::skip(),
    ]);

    table.print(
        TableOpts::new()
            .border(BorderType::Single)
            .header_row(true)
            .title("Spans"),
    );
}

/// Stripes, a header column and markup cells; per-row background highlight.
fn print_styled_table() {
    let mut table = Table::new();
    table
        .column("Check", ColOpts::new())
        .column("Result", ColOpts::new())
        .column("Duration", ColOpts::new().align(Align::Right));

    table.row([Cell::new("build"), Cell::markup("[green]✔ passed[/]"),
               Cell::new("41 s")]);
    table.row([Cell::new("lint"), Cell::markup("[green]✔ passed[/]"),
               Cell::new("3 s")]);
    // Per-row background highlights the failing check.
    table.row_bg(
        [Cell::new("fuzz"), Cell::markup("[red]✖ failed[/]"),
         Cell::new("122 s")],
        Color::rgb(60, 30, 30),
    );

    table.print(
        TableOpts::new()
            .border(BorderType::Rounded)
            .header_col(true)
            .striped(true),
    );
}
